ygot/pathstrings_test.go

```diff
@@ -0,0 +1,352 @@
+				{Name: "interfaces"},
+				{Name: "interface", Key: map[string]string{"name": `[\]`}},
+			},
+		},
+	}, {
+		name:                "forward slash in key which does not need to be escaped ",
+		in:                  `/interfaces/interface[name=\/foo]/state`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"interfaces", `interface[name=\/foo]`, "state"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "interfaces"},
+				{Name: "interface", Key: map[string]string{"name": `/foo`}},
+				{Name: "state"},
+			},
+		},
+	}, {
+		name:                "escaped forward slash in an element name",
+		in:                  `/interfaces/inter\/face[name=foo]`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"interfaces", "inter/face[name=foo]"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "interfaces"},
+				{Name: "inter/face", Key: map[string]string{"name": "foo"}},
+			},
+		},
+	}, {
+		name:                `single-level wildcard`,
+		in:                  `/interfaces/interface/*/state`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"interfaces", "interface", "*", "state"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "interfaces"},
+				{Name: "interface"},
+				{Name: "*"},
+				{Name: "state"},
+			},
+		},
+	}, {
+		name:                "multi-level wildcard",
+		in:                  `/interfaces/.../state`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"interfaces", "...", "state"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "interfaces"},
+				{Name: "..."},
+				{Name: "state"},
+			},
+		},
+	}, {
+		name:                "path with escaped backslash in an element",
+		in:                  `/foo/bar\\\/baz/hat`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"foo", `bar\/baz`, "hat"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "foo"},
+				{Name: `bar/baz`},
+				{Name: "hat"},
+			},
+		},
+	}, {
+		name:                "path with escaped backslash in a key",
+		in:                  `/foo/bar[baz\\foo=hat]`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"foo", `bar[baz\\foo=hat]`}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "foo"},
+				{Name: "bar", Key: map[string]string{`baz\foo`: "hat"}},
+			},
+		},
+	}, {
+		name:                "additional equals within the key, unescaped",
+		in:                  `/foo/bar[baz==bat]`,
+		wantStringSlicePath: &gnmipb.Path{Element: []string{"foo", "bar[baz==bat]"}},
+		wantStructuredPath: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{
+				{Name: "foo"},
+				{Name: "bar", Key: map[string]string{"baz": "=bat"}},
+			},
+		},
+	}, {
+		name:              "error - unescaped ] within a key value",
+		in:                `/foo/bar[baz=]bat]`,
+		wantSliceErr:      "error parsing path /foo/bar[baz=]bat]: received null value for key baz of element bar",
```

Comment:
adding specific check in the code would lead to more meaningful error message here.

Reply:
What is the specific check here though?

1) We are in a key, so ] is valid unescaped because it means the end of the key.
2) We are in the value of a key, where we expect the ] character.
3) We might have additional characters remaining if the next character is [.

So, we need to refactor entirely here again to be able to do this since we need to check that the next character is not a [. Is this worth doing just for a slightly better error message?